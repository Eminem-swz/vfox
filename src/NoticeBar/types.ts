export type Mode = 'default' | 'clickable' | 'closable'
