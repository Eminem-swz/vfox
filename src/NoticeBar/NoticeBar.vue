<template>
  <div
    class="fx-notice-bar"
    :class="[typeClassName]"
    :style="styles"
    v-show="visible2"
  >
    <div v-if="leftIcon" class="fx-notice-bar_left-icon">
      <icon :icon="leftIcon" :style="iconStyles" />
    </div>
    <div class="fx-notice-bar_content">
      <div
        ref="content"
        class="fx-notice-bar_content-inner"
        :class="{ marquee: !!marquee }"
        :style="contentStyles"
      >
        <slot>
          {{ title }}
        </slot>
      </div>
    </div>
    <div
      v-if="rightIcon2"
      class="fx-notice-bar_right-icon"
      @click="onRightIconClick"
    >
      <icon :icon="rightIcon2" :style="iconStyles" />
    </div>
  </div>
</template>

<script lang="ts">
import {
  defineComponent,
  computed,
  onMounted,
  onBeforeUnmount,
  inject,
  ref,
  watch,
  PropType
} from 'vue'
import Icon from '@/Icon'
import {
  createEnumsValidator,
  getEnumsValue,
  iconValidator
} from '@/utils/validator'
import  type {  StyleObject } from '@/utils/types'
import { StateTypes, STATE_TYPES } from '@/utils/constants'

const modeMaps = new Map([
  ['default', ''],
  ['clickable', 'RightOutlined'],
  ['closable', 'CloseOutlined']
])

export default defineComponent({
  name: 'fx-notice-bar',
  components: { Icon },
  inject: {
    appNotify: {
      default: null
    }
  },
  props: {
    // 是否显示
    visible: {
      type: Boolean,
      default: true
    },
    title: {
      type: String,
      default: ''
    },
    // 通知栏模式
    mode: {
      type: String as PropType<'default' | 'clickable' | 'closable'>,
      validator: (val: string) => {
        return modeMaps.get(val) != null
      },
      default: 'default'
    },
    // 左侧图标名称
    leftIcon: {
      validator: iconValidator,
      default: null
    },
    // 右边侧图标名称
    rightIcon: {
      validator: iconValidator,
      default: null
    },
    // 背景颜色
    backgroundColor: {
      type: String,
      default: null
    },
    // 字体颜色
    color: {
      type: String,
      default: null
    },
    // 是否采用跑马灯显示
    marquee: {
      type: Boolean,
      default: false
    },
    type: {
      type: String as PropType<StateTypes>,
      validator: createEnumsValidator(STATE_TYPES),
      default: null
    }
  },
  emits: ['update:visible', 'show', 'hide', 'close-click'],
  setup(props, ctx) {
    const { emit } = ctx
    const appNotify = inject('appNotify', null)
    const visible2 = ref(!!props.visible)
    const marqueeX = ref(0)
    const marqueeDuration = ref(0)
    const content = ref<HTMLElement>()
    let marqueeTimer: number

    function marqueeStep(x: number, pW: number) {
      marqueeX.value = pW
      marqueeDuration.value = 0

      marqueeTimer = window.setTimeout(() => {
        marqueeX.value = -x
        marqueeDuration.value = x / 30

        marqueeTimer = window.setTimeout(() => {
          marqueeStep(x, pW)
        }, (x / 28) * 1000)
      }, 17)
    }

    function stopMarquee() {
      clearTimeout(marqueeTimer)

      marqueeX.value = 0
      marqueeDuration.value = 0
    }

    function startMarquee() {
      stopMarquee()

      const $content = content.value as HTMLElement

      const w = $content.offsetWidth
      const pW = ($content.parentElement as HTMLElement).offsetWidth

      if (w > pW) {
        marqueeStep(w, pW)
      }
    }

    function resetMarquee() {
      props.marquee ? startMarquee() : stopMarquee()
    }

    function show() {
      if (!visible2.value) {
        emit('update:visible', true)
        visible2.value = true

        emit('show', {})
      }
    }
    function hide() {
      if (visible2.value) {
        emit('update:visible', false)
        visible2.value = false

        emit('hide', {})
      }
    }

    function onRightIconClick() {
      if (props.mode === 'closable') {
        if (!appNotify) {
          hide()
        }

        emit('close-click', {})
      }
    }

    onMounted(() => props.marquee && startMarquee())

    onBeforeUnmount(() => stopMarquee())

    watch([() => props.marquee, () => props.title], () => {
      resetMarquee()
    })

    const rightIcon2 = computed(() => {
      return props.rightIcon || modeMaps.get(props.mode) || null
    })

    const styles = computed(() => {
      const obj: StyleObject = {}

      props.backgroundColor && (obj.backgroundColor = props.backgroundColor)
      props.color && (obj.color = props.color)

      return obj
    })

    const iconStyles = computed(() => {
      const obj: StyleObject = {}

      props.color && (obj.fill = props.color)

      return obj
    })

    const contentStyles = computed(() => {
      const obj: StyleObject = {}

      marqueeX.value !== 0 &&
        (obj.transform = `translateX(${marqueeX.value}px)`)
      marqueeDuration.value > 0 &&
        (obj.transitionDuration = `${marqueeDuration.value}s`)

      return obj
    })

    // 计算属性的 getter
    const typeClassName = computed(() => {
      return 'type--' + getEnumsValue(STATE_TYPES, props.type)
    })

    watch(() => props.visible, val => {
      visible2.value = !!val
    })

    return {
      show,
      hide,
      content,
      visible2,
      rightIcon2,
      styles,
      iconStyles,
      contentStyles,
      typeClassName,
      onRightIconClick
    }
  }
})
</script>
