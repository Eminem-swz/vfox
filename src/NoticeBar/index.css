.fx-notice-bar {
  --notice-bar-color: #db8f02;
  --notice-bar-background-color: #fffae0;
  position: relative;
  display: flex;
  padding: 10px 0 10px 16px;
  min-height: 40px;
  align-items: center;
  background-color: var(--notice-bar-background-color);
  color: var(--notice-bar-color);
  line-height: 20px;
  font-size: 14px;
  box-sizing: border-box;
  flex-grow: 1;
}

.fx-notice-bar.type--success {
  --notice-bar-color: #52c41a;
  --notice-bar-background-color: #f0ffe0;
}

.fx-notice-bar.type--primary {
  --notice-bar-color: #1890ff;
  --notice-bar-background-color: #e6f7ff;
}

.fx-notice-bar.type--danger {
  --notice-bar-color: #ff4d4f;
  --notice-bar-background-color: #ffe7e6;
}

.fx-notice-bar_left-icon, .fx-notice-bar_right-icon {
  display: flex;
  align-items: center;
  height: 100%;
  margin-right: 8px;
  flex-shrink: 0;
}

.fx-notice-bar_left-icon .fx-icon, .fx-notice-bar_right-icon .fx-icon {
  --size: 17px;
  --color: var(--notice-bar-color);
}

.fx-notice-bar_right-icon {
  padding: 0 16px 0 8px;
  margin: -10px 0;
}

.fx-notice-bar_content {
  flex-grow: 1;
  overflow: hidden;
  margin-right: 16px;
  position: relative;
  min-height: 20px;
  text-align: left;
}

.fx-notice-bar_content + .fx-notice-bar_right-icon {
  margin-left: -16px;
}

.fx-notice-bar_content-inner.marquee {
  position: absolute;
  white-space: nowrap;
  transition-timing-function: linear;
}
