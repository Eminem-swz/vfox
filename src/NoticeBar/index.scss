@import '../style/var.scss';

.fx-notice-bar {
  // --notice-bar-color: #{$warning-color};
  --notice-bar-color: #db8f02;
  --notice-bar-background-color: #{$warning-background-color};

  position: relative;
  display: flex;
  padding: 10px 0 10px 16px;
  min-height: 40px;
  align-items: center;
  background-color: var(--notice-bar-background-color);
  color: var(--notice-bar-color);
  line-height: 20px;
  font-size: 14px;
  box-sizing: border-box;
  flex-grow: 1;

  &.type--success {
    --notice-bar-color: #{$success-color};
    --notice-bar-background-color: #{$success-background-color};
  }

  &.type--primary {
    --notice-bar-color: #{$primary-color};
    --notice-bar-background-color: #{$primary-background-color};
  }

  &.type--danger {
    --notice-bar-color: #{$danger-color};
    --notice-bar-background-color: #{$danger-background-color};
  }

  &_left-icon,
  &_right-icon {
    display: flex;
    align-items: center;
    height: 100%;
    margin-right: 8px;
    flex-shrink: 0;

    .fx-icon {
      --size: 14px;
      --color: var(--notice-bar-color);
    }
  }

  &_right-icon {
    padding: 0 16px 0 8px;
    margin: -10px 0;
  }

  &_content {
    flex-grow: 1;
    overflow: hidden;
    margin-right: 16px;
    position: relative;
    min-height: 20px;
    text-align: left;

    + .fx-notice-bar_right-icon {
      margin-left: -16px;
    }
  }

  &_content-inner {
    &.marquee {
      position: absolute;
      white-space: nowrap;
      transition-timing-function: linear;
    }
  }
}
