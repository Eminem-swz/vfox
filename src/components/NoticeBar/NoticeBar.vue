<template>
  <div
    :class="[prefix + '-notice-bar', typeClassName]"
    :style="styles"
    v-show="visible2"
    @click="onClick"
  >
    <div v-if="leftIcon" :class="[prefix + '-notice-bar_left-icon']">
      <icon :class-name="leftIcon" :style="iconStyle"/>
    </div>
    <div :class="[prefix + '-notice-bar_content']">
      <div
        ref="content"
        :class="[prefix + '-notice-bar_content-inner', { marquee: !!marquee }]"
        :style="contentStyles"
      >
        {{ title }}
      </div>
    </div>
    <div
      v-if="rightIcon2"
      :class="[prefix + '-notice-bar_right-icon']"
      @click="onRightIconClick"
    >
      <icon :class-name="rightIcon2" :style="iconStyle"/>
    </div>
  </div>
</template>

<script>
import Icon from '../Icon'
import { SDKKey } from '../../config'
import { inArray } from '../../helpers/util'

const TYPE_NAMES = ['primary', 'success', 'warning', 'danger']

const MODE_MAPS = {
  default: '',
  clickable: 'RightOutlined',
  closable: 'CloseOutlined'
}

export default {
  name: SDKKey + '-notice-bar',
  components: { Icon },
  inject: {
    appNotify: {
      default: null
    }
  },
  props: {
    // 是否显示
    visible: {
      type: Boolean,
      default: true
    },
    title: {
      type: String,
      required: true,
      default: ''
    },
    // 通知栏模式
    mode: {
      validator(val) {
        return MODE_MAPS[val] != null
      },
      default: 'default'
    },
    // 左侧图标名称
    leftIcon: {
      type: String,
      default: null
    },
    // 右边侧图标名称
    rightIcon: {
      type: String,
      default: null
    },
    // 背景颜色
    backgroundColor: {
      type: String,
      default: null
    },
    // 字体颜色
    color: {
      type: String,
      default: null
    },
    // 是否采用跑马灯显示
    marquee: {
      type: Boolean,
      default: false
    },
    type: {
      validator(val) {
        return inArray(val, TYPE_NAMES)
      },
      default: TYPE_NAMES[2]
    }
  },
  data() {
    return { prefix: SDKKey, marqueeX: 0, marqueeDuration: 0, visible2: true }
  },
  watch: {
    marquee() {
      this.resetMarquee()
    },
    title() {
      this.resetMarquee()
    },
    visible: {
      immediate: true,
      handler(val) {
        this.visible2 = !!val
      }
    }
  },
  computed: {
    rightIcon2() {
      if (this.rightIcon) {
        return this.rightIcon
      }

      if (this.mode) {
        return MODE_MAPS[this.mode] || null
      }

      return null
    },
    styles() {
      const obj = {}

      if (this.backgroundColor) obj.backgroundColor = this.backgroundColor
      if (this.color) obj.color = this.color

      return obj
    },
    iconStyle() {
      const obj = {}

      if (this.color) obj.fill = this.color

      return obj
    },
    contentStyles() {
      const obj = {}

      if (this.marqueeX !== 0) obj.transform = `translateX(${this.marqueeX}px)`
      if (this.marqueeDuration > 0)
        obj.transitionDuration = `${this.marqueeDuration}s`

      return obj
    },
    // 计算属性的 getter
    typeClassName() {
      return (
        'type--' + (inArray(this.type, TYPE_NAMES) ? this.type : TYPE_NAMES[0])
      )
    }
  },
  mounted() {
    if (this.marquee) {
      this.startMarquee()
    }
  },
  destroyed() {
    this.stopMarquee()
  },
  methods: {
    resetMarquee() {
      if (this.marquee) {
        this.startMarquee()
      } else {
        this.stopMarquee()
      }
    },

    startMarquee() {
      this.stopMarquee()

      const $content = this.$refs.content

      if ($content.offsetWidth > $content.parentNode.offsetWidth) {
        this.marqueeStep($content.offsetWidth)
      }
    },

    stopMarquee() {
      clearTimeout(this.marqueeTimer)

      this.marqueeX = 0
      this.marqueeDuration = 0
    },

    marqueeStep(x) {
      this.marqueeX = x
      this.marqueeDuration = 0

      this.marqueeTimer = setTimeout(() => {
        this.marqueeX = -x
        this.marqueeDuration = x / 30

        this.marqueeTimer = setTimeout(() => {
          this.marqueeStep(x)
        }, (x / 30) * 1000)
      }, 17)
    },

    show() {
      if (!this.visible2) {
        this.$emit('update:visible', true)
        this.visible2 = true

        this.$emit('show', {})
      }
    },
    hide() {
      if (this.visible2) {
        this.$emit('update:visible', false)
        this.visible2 = false

        this.$emit('hide', {})
      }
    },
    onRightIconClick() {
      if (this.mode === 'closable') {
        if (!this.appNotify) {
          this.hide()
        }

        this.$emit('close-click', {})
      }
    },
    onClick(e) {
      this.$emit(e.type, e)
    }
  }
}
</script>

<style lang="scss">
@import '../component.module.scss';

.#{$prefix}-notice-bar {
  // --notice-bar-color: #{$warning-color};
  --notice-bar-color: #db8f02;
  --notice-bar-background-color: #{$warning-background-color};

  position: relative;
  display: flex;
  padding: 10px 0 10px 16px;
  min-height: 40px;
  align-items: center;
  background-color: var(--notice-bar-background-color);
  color: var(--notice-bar-color);
  line-height: 20px;
  font-size: 14px;
  box-sizing: border-box;
  flex-grow: 1;

  &.type--success {
    --notice-bar-color: #{$success-color};
    --notice-bar-background-color: #{$success-background-color};
  }

  &.type--primary {
    --notice-bar-color: #{$primary-color};
    --notice-bar-background-color: #{$primary-background-color};
  }

  &.type--danger {
    --notice-bar-color: #{$danger-color};
    --notice-bar-background-color: #{$danger-background-color};
  }

  &_left-icon,
  &_right-icon {
    display: flex;
    align-items: center;
    height: 100%;
    margin-right: 8px;
    flex-shrink: 0;

    .#{$prefix}-icon {
      --size: 14px;
      --color: var(--notice-bar-color);
    }
  }

  &_right-icon {
    padding: 0 16px 0 8px;
    margin: -10px 0;
  }

  &_content {
    flex-grow: 1;
    overflow: hidden;
    margin-right: 16px;
    position: relative;
    min-height: 20px;

    + .#{$prefix}-notice-bar_right-icon {
      margin-left: -16px;
    }
  }

  &_content-inner {
    &.marquee {
      position: absolute;
      white-space: nowrap;
      transition-timing-function: linear;
    }
  }
}
</style>
