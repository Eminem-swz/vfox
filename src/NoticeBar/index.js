import NoticeBar from './NoticeBar.vue'

NoticeBar.install = function(Vue) {
  Vue.component(NoticeBar.name, NoticeBar)
}

export default NoticeBar
