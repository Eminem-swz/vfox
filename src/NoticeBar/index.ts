import { App } from 'vue'
import { SFCWithInstall } from '../utils/types'
import NoticeBar from './NoticeBar.vue'

const _NoticeBar: SFCWithInstall<typeof NoticeBar> = Object.assign(NoticeBar, {
  install: function(app: App) {
    app.component(NoticeBar.name, NoticeBar)
  }
})

export default _NoticeBar
