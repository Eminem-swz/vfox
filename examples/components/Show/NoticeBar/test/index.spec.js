/* global describe,test,expect */

import { mount } from '@vue/test-utils'
import NoticeBar from '@/components/NoticeBar'

describe('NoticeBar', () => {
  test('should render default correctly', () => {
    const wrapper = mount(NoticeBar, {
      propsData: {
        title: 'title'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render prop leftIcon="SoundOutlined" correctly', () => {
    const wrapper = mount(NoticeBar, {
      propsData: {
        title: 'title',
        leftIcon: 'SoundOutlined'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render prop mode="closable" correctly', () => {
    const wrapper = mount(NoticeBar, {
      propsData: {
        title: 'title',
        mode: 'closable'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should render prop mode="clickable" correctly', () => {
    const wrapper = mount(NoticeBar, {
      propsData: {
        title: 'title',
        mode: 'clickable'
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
