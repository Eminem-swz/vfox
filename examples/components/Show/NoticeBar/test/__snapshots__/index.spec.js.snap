// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoticeBar should render default correctly 1`] = `
<div class="fx-notice-bar type--warning">
  <!---->
  <div class="fx-notice-bar_content">
    <div class="fx-notice-bar_content-inner">
      title
    </div>
  </div>
  <!---->
</div>
`;

exports[`NoticeBar should render prop leftIcon="SoundOutlined" correctly 1`] = `
<div class="fx-notice-bar type--warning">
  <div class="fx-notice-bar_left-icon"><svg aria-hidden="true" class="fx-icon">
      <use xlink:href="#icon-SoundOutlined"></use>
    </svg></div>
  <div class="fx-notice-bar_content">
    <div class="fx-notice-bar_content-inner">
      title
    </div>
  </div>
  <!---->
</div>
`;

exports[`NoticeBar should render prop mode="clickable" correctly 1`] = `
<div class="fx-notice-bar type--warning">
  <!---->
  <div class="fx-notice-bar_content">
    <div class="fx-notice-bar_content-inner">
      title
    </div>
  </div>
  <div class="fx-notice-bar_right-icon"><svg aria-hidden="true" class="fx-icon">
      <use xlink:href="#icon-RightOutlined"></use>
    </svg></div>
</div>
`;

exports[`NoticeBar should render prop mode="closable" correctly 1`] = `
<div class="fx-notice-bar type--warning">
  <!---->
  <div class="fx-notice-bar_content">
    <div class="fx-notice-bar_content-inner">
      title
    </div>
  </div>
  <div class="fx-notice-bar_right-icon"><svg aria-hidden="true" class="fx-icon">
      <use xlink:href="#icon-CloseOutlined"></use>
    </svg></div>
</div>
`;
